File: Sliding Window/239. Sliding Window Maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        output = []
        queue = collections.deque()
        l = r = 0
        
        for i in range(len(nums)):
            r = i
            
            while queue and nums[queue[-1]] < nums[r]:
                queue.pop()
            queue.append(r)
            
            
            if l > queue[0]:
                queue.popleft()
                
            if (r+1) >= k:
                output.append(nums[queue[0]])
                l+=1

        return output

File: Sliding Window/239. Sliding Window Maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> res;
        deque<pair<int, int>>q;
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            if (!q.empty() && q.front().second < i - k + 1) {
                q.pop_front();
            }
            while (!q.empty() && q.back().first < nums[i]) {
                q.pop_back();
            }
            q.push_back({nums[i], i});
            if (i - k + 1 >= 0) res.push_back(q.front().first);
        }

        return res;
    }
};